File: content_tool/publishers/wp_factory.py
"""Resolve the WordPress publish target for a run from its voice (persona).

Phase 1 supports multiple WordPress instances. A voice maps to a
``publish_targets`` row via ``personas.publish_target_id``; the row carries an
``auth_ref`` env-var prefix from which the base URL + credentials are read at
publish time (``{auth_ref}_BASE_URL`` / ``_USERNAME`` / ``_APP_PASSWORD``).
Secrets never live in the database.

A voice with no assigned target (NULL FK) — and any pre-existing run — resolves
to the process-default client built from the legacy ``WP_*`` settings, so
behaviour is unchanged until a voice is explicitly pointed elsewhere.
"""

import os
from collections.abc import Mapping
from dataclasses import dataclass

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from content_tool.db.persona_model import Persona
from content_tool.db.publish_target_model import PublishTarget
from content_tool.wordpress.client import WordPressClient

# Phase 1 ships WordPress only; widen when Ghost lands (mirrors the DB CHECK).
_SUPPORTED_KINDS = frozenset({"wordpress"})


@dataclass(frozen=True)
class ResolvedTarget:
    """The WordPress client a run should publish through, plus its display label.

    ``base_url`` / ``username`` / ``app_password`` are populated only for
    non-default targets so the caller can build a matching SEO-plugin resolver
    against that instance. For the default target the caller keeps using its
    existing (process-level) SEO resolver.
    """

    client: WordPressClient | None
    label: str
    is_default: bool
    base_url: str | None = None
    username: str | None = None
    app_password: str | None = None


def _require_env(env: Mapping[str, str], key: str) -> str:
    value = env.get(key)
    if not value:
        raise OSError(
            f"publish target requires env var {key!r}, which is not set"
        )
    return value


def build_target_client(
    target: PublishTarget,
    *,
    timeout: float,
    env: Mapping[str, str],
) -> ResolvedTarget:
    """Build a :class:`ResolvedTarget` from a publish-target row + env creds.

    Pure (no DB / no process env unless ``env`` is ``os.environ``). Raises
    ``ValueError`` for an archived or unsupported-kind target and
    ``EnvironmentError`` when a required credential env var is absent.
    """
    if target.is_archived:
        raise ValueError(
            f"publish target {target.name!r} is archived and cannot be used"
        )
    if target.kind not in _SUPPORTED_KINDS:
        raise ValueError(
            f"unsupported publish target kind {target.kind!r} "
            f"(supported: {sorted(_SUPPORTED_KINDS)})"
        )
    ref = target.auth_ref
    base_url = _require_env(env, f"{ref}_BASE_URL")
    username = _require_env(env, f"{ref}_USERNAME")
    app_password = _require_env(env, f"{ref}_APP_PASSWORD")
    client = WordPressClient(
        base_url, username=username, app_password=app_password, timeout=timeout
    )
    return ResolvedTarget(
        client=client,
        label=target.name,
        is_default=False,
        base_url=base_url,
        username=username,
        app_password=app_password,
    )


async def resolve_wp_target(
    *,
    session: AsyncSession,
    persona_slug: str | None,
    default_client: WordPressClient | None,
    default_label: str,
    timeout: float = 15.0,  # noqa: ASYNC109 — WP client config, not an async cancel budget
    env: Mapping[str, str] | None = None,
) -> ResolvedTarget:
    """Resolve the publish target for a voice.

    NULL/unknown voice or unassigned target → the process-default client.
    An assigned, active target → a client built from its ``auth_ref`` env creds.
    Raises ``ValueError`` if the voice references a missing/archived target and
    ``EnvironmentError`` if its credential env vars are absent.
    """
    resolved_env: Mapping[str, str] = env if env is not None else os.environ
    default = ResolvedTarget(
        client=default_client, label=default_label, is_default=True
    )
    if not persona_slug:
        return default

    persona = (
        await session.execute(select(Persona).where(Persona.slug == persona_slug))
    ).scalar_one_or_none()
    if persona is None or persona.publish_target_id is None:
        return default

    target = await session.get(PublishTarget, persona.publish_target_id)
    if target is None:
        raise ValueError(
            f"publish target {persona.publish_target_id} referenced by voice "
            f"{persona_slug!r} was not found"
        )
    return build_target_client(target, timeout=timeout, env=resolved_env)

File: deploy/cloudflare-workers/src/publishers/wp_factory.ts
import type { Env } from "../index";
import type { getSql } from "../db/client";
import type { PublishTargetRow } from "../db/schema";
import { getPublishTargetForVoice } from "../db/publish_targets";

// Supported publish-target kinds (mirrors the DB CHECK in
// 20260618000000_publish_targets_ghost.sql and Python wp_factory.py).
const SUPPORTED_KINDS = new Set(["wordpress", "ghost"]);

export interface ResolvedTarget {
  // Env-var prefix for the target's credentials, or null for the default.
  // NOT a secret — safe to persist as a Workflow step result.
  authRef: string | null;
  label: string;
  isDefault: boolean;
  // "wordpress" (default) or "ghost" — selects the publisher at publish time.
  kind: string;
}

/**
 * Turn a publish-target row (or null for an unassigned voice) into a resolved
 * target descriptor. Pure — no DB, no env. Throws for an archived or
 * unsupported-kind target.
 */
export function targetFromRow(
  row: PublishTargetRow | null,
  defaultLabel: string,
): ResolvedTarget {
  if (row === null) {
    return { authRef: null, label: defaultLabel, isDefault: true, kind: "wordpress" };
  }
  if (row.is_archived) {
    throw new Error(`publish target '${row.name}' is archived and cannot be used`);
  }
  if (!SUPPORTED_KINDS.has(row.kind)) {
    throw new Error(`unsupported publish target kind '${row.kind}'`);
  }
  return { authRef: row.auth_ref, label: row.name, isDefault: false, kind: row.kind };
}

/**
 * Resolve the publish target for a voice. NULL/unknown voice → the default
 * (legacy WP env). Returns only non-secret descriptors, so the result is safe
 * to persist as a durable Workflow step result.
 */
export async function resolvePublishTarget(
  sql: ReturnType<typeof getSql>,
  personaSlug: string,
  defaultLabel: string,
): Promise<ResolvedTarget> {
  const row = await getPublishTargetForVoice(sql, personaSlug);
  return targetFromRow(row, defaultLabel);
}

/**
 * Build an Env whose WP_* credentials point at the resolved target, reading them
 * from the process env under the target's auth_ref prefix
 * ({ref}_BASE_URL / _USERNAME / _APP_PASSWORD). The default target returns env
 * unchanged. Credentials are read here (in the running Worker), never persisted.
 * Throws when a required credential env var is absent.
 */
export function buildTargetEnv(env: Env, target: ResolvedTarget): Env {
  // Ghost targets carry no WP_* creds; the GhostPublisher reads its own
  // ({ref}_API_URL / {ref}_ADMIN_API_KEY) via buildGhostCreds at publish time.
  if (target.isDefault || target.authRef === null || target.kind !== "wordpress") {
    return env;
  }
  const ref = target.authRef;
  const get = (key: string): string => {
    const value = (env as unknown as Record<string, string | undefined>)[key];
    if (!value) {
      throw new Error(`publish target requires env var ${key}, which is not set`);
    }
    return value;
  };
  return {
    ...env,
    WP_BASE_URL: get(`${ref}_BASE_URL`),
    WP_USERNAME: get(`${ref}_USERNAME`),
    WP_APP_PASSWORD: get(`${ref}_APP_PASSWORD`),
  };
}
